addons/roe/roe.lua

```diff
@@ -0,0 +1,280 @@
+      -- derived from this software without specific prior written permission.
+
+-- THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+-- ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+-- WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+-- DISCLAIMED. IN NO EVENT SHALL Cair BE LIABLE FOR ANY
+-- DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+-- (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+-- LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+-- ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+-- (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+-- SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+_addon = {}
+_addon.name = 'ROE'
+_addon.version = '1.0'
+_addon.author = "Cair"
+_addon.commands = {'roe'}
+
+packets = require('packets')
+config = require('config')
+require('logger')
+
+
+local defaults = T{
+    profiles = T{
+        default = S{},
+    },
+    blacklist = S{},
+    clear = true,
+    clearprogress = false,
+    clearall = false,
+}
+
+settings = config.load(defaults)
+
+_roe = T{
+    active = T{},
+    complete = T{}    
+}
+
+
+local function cancel_roe(id)
+    id = tonumber(id)
+    
+    if not id then return end
+    
+    if settings.blacklist[id] then return end
+    
+    local p = packets.new('outgoing', 0x10d, {['RoE Quest'] = id })
+    packets.inject(p)
+end
+
+local function accept_roe(id)
+    id = tonumber(id)
+    
+    if not id or _roe.complete[id] or _roe.active[id] then return end
+    
+    local p = packets.new('outgoing', 0x10c, {['RoE Quest'] = id })
+    packets.inject(p)
+end
+
+local function eval(...)
+    assert(loadstring(table.concat({...}, ' ')))()
+end
+
+local function save(name)
+    if not type(name) == "string" then
+        error('`save` : specify a profile name')
+        return
+    end
+    
+    name = name:lower()
+
+    settings.profiles[name] = S(_roe.active:keyset())
+    settings:save('global')
+    notice('saved %d objectives to the profile %s':format(_roe.active:length(), name))
+end
+
+
+local function list()
+    notice('You have saved the following profiles: ')
+    notice(settings.profiles:keyset())
+end
+
+local function set(name)
+    if not type(name) == "string" then
+        error('`set` : specify a profile name')
+        return
+    end
+    
+    name = name:lower()
+    
+    if not settings.profiles[name] then
+        error('`set` : the profile \'%s\' does not exist':format(name))
+        return
+    end
+    
+    if type(settings.profiles[name]) == "string" then
+        settings.profiles[name] = S(settings.profiles[name]:split(','):map(tonumber))
+    end
+    
+
+    
+    local needed_quests = settings.profiles[name]:diff(_roe.active:keyset())   
+    local available_slots = 30 - _roe.active:length()

```

Comment:
Magic number! Is that a fixed constant? Maybe put it somewhere near the top for easy adjustment, if that can be ever adjusted.

I just saw below it was the length of the packet payload, so we could derive the number from the `fields.lua` file. That info should be available and exposed through the `packets` API.